brave-kafka/src/main/java/smartthings/brave/kafka/consumers/TracingConsumerInterceptor.java

```diff
@@ -0,0 +1,128 @@
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package smartthings.brave.kafka.consumers;
+
+import brave.Tracer;
+import brave.propagation.TraceContext;
+import brave.propagation.TraceContextOrSamplingFlags;
+import com.google.protobuf.InvalidProtocolBufferException;
+import org.apache.kafka.clients.ClientUtils;
+import org.apache.kafka.clients.consumer.ConsumerInterceptor;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.config.ConfigException;
+import smartthings.brave.kafka.EnvelopeProtos;
+import zipkin.Constants;
+import zipkin.Endpoint;
+
+import java.math.BigInteger;
+import java.net.InetSocketAddress;
+import java.util.*;
+
+public class TracingConsumerInterceptor<K> implements ConsumerInterceptor<K, byte[]> {
+
+  private Tracer tracer;
+  private Endpoint kafkaEndpoint;
+
+  // TODO: how to inject dependencies without using config
+//  private final SpanNameProvider<K> nameProvider;
+
+  public TracingConsumerInterceptor() {}
+
+  private final TraceContext.Extractor<EnvelopeProtos.Envelope> extractor = new TraceContext.Extractor<EnvelopeProtos.Envelope>() {
+    @Override
+    public TraceContextOrSamplingFlags extract(EnvelopeProtos.Envelope envelope) {
+      TraceContext.Builder builder = TraceContext.newBuilder()
+        .traceIdHigh(envelope.getTraceIdHigh())
+        .traceId(envelope.getTraceId())
+        .spanId(envelope.getSpanId())
+        .sampled(envelope.getSampled())
+        .shared(envelope.getShared());
+      if (envelope.getParentId() != 0) {
+        builder.parentId(envelope.getParentId());
+      }
+      return TraceContextOrSamplingFlags.create(builder);
+    }
+  };
+
+
+  @Override
+  public ConsumerRecords<K, byte[]> onConsume(ConsumerRecords<K, byte[]> records) {
+    Map<TopicPartition, List<ConsumerRecord<K, byte[]>>> unenvelopedRecords = new HashMap<>(records.count());
+    records.forEach(record -> {
+      try {
+        EnvelopeProtos.Envelope envelope = EnvelopeProtos.Envelope.parseFrom(record.value());
+        TraceContextOrSamplingFlags traceContextOrSamplingFlags = extractor.extract(envelope);
+        TraceContext ctx = traceContextOrSamplingFlags.context();
+        (ctx != null ? tracer.joinSpan(ctx) : tracer.newTrace(traceContextOrSamplingFlags.samplingFlags()))
+          .name("Receive " + record.topic() + "-" + record.partition())
```

Comment:
I cannot use non-default constructor for these interceptor classes so I am just defaulting the span name to topic and partition. If you don't mind passing dependencies in the config I will add the span name provider there as well.